File: MarioKart/MarioKart/Scenes/GameScene.m
//
//  GameScene.m
//  MarioKart
//
//  Created by Txai Wieser on 22/11/15.
//  Copyright © 2015 Txai Wieser. All rights reserved.
//

#import "GameScene.h"
#import "Mushroom.h"
#import "RWTCube.h"
#import "QMarkBox.h"

@interface GameScene ()

@property (strong, nonatomic) TWGLNode *qmark;

@end

@implementation GameScene {
    TWGLNode *xCube;
    TWGLNode *yCube;
    TWGLNode *zCube;
}

- (instancetype)initWithShader:(TWGLShadersReference *)shader {
    self = [super initWithShader:shader];
    if (self) {
        _camera = [[TWGLCamera alloc] initWithShader:shader];
        _carNode = [[CarNode alloc] initWithShader:shader];
        _qmark = [[QMarkBox alloc] initWithShader:shader];
        [self.children addObject:_carNode];
        [self.children addObject:_camera];
        [self.children addObject:_qmark];
        
        xCube = [[Mushroom alloc] initWithShader:shader];
        xCube.position = GLKVector3Make(20, 0, 0);
        [self.children addObject:xCube];
        
        yCube = [[Mushroom alloc] initWithShader:shader];
        yCube.position = GLKVector3Make(0, 20, 0);
        [self.children addObject:yCube];
        
        zCube = [[Mushroom alloc] initWithShader:shader];
        zCube.position = GLKVector3Make(0, 0, 20);
        [self.children addObject:zCube];

    }
    return self;
}

- (void)render {
    GLKMatrix4 viewMatrix = GLKMatrix4Multiply(self.carNode.modelMatrix, self.camera.modelMatrix);
    viewMatrix = GLKMatrix4Invert(viewMatrix, nil);
    [super renderWithParentModelViewMatrix:viewMatrix];
}

- (void)updateWithDelta:(NSTimeInterval)dt {
    [super updateWithDelta:dt];
}

@end

File: MarioKart/MarioKart/Scenes/GameScene.swift
//
//  GameScene.swift
//  MarioKart
//
//  Created by Txai Wieser on 05/11/15.
//  Copyright © 2015 Txai Wieser. All rights reserved.
//

import GLKit
import OpenGLES

class GameScene: TWGLScene {
    var floor:TWGLNode!
    
    var xCube:TWGLNode!
    var yCube:TWGLNode!
    var zCube:TWGLNode!

    var mushroom:TWGLNode!
    var carNode:CarNode!
    
    override init(shader: TWGLShaderReference) {
        super.init(shader: shader)

//        floor = FloorNode(shader: shader)
//        children.append(floor)

        xCube = CubeNode(shader: shader)
        xCube.matColor = (1.0, 0.0, 0.0, 1.0)
        xCube.position.x = 20
        children.append(xCube)

        yCube = CubeNode(shader: shader)
        yCube.matColor = (0.0, 1.0, 0.0, 1.0)
        yCube.position.y = 20
        children.append(yCube)
        
        zCube = CubeNode(shader: shader)
        zCube.matColor = (0.0, 0.0, 1.0, 1.0)
        zCube.position.z = 20
        children.append(zCube)

        mushroom = MushroomNode(shader: shader)
        children.append(mushroom)
        
        carNode = CarNode(shader: shader)
        carNode.children.append(camera)
        camera.position.y = 5
        camera.rotation.x = -Float(M_PI)/6
        camera.position.z = 10
        children.append(carNode)
    }
    
    func render() {
        var viewMatrix = GLKMatrix4Multiply(carNode.modelMatrix(), camera.modelMatrix())
        viewMatrix = GLKMatrix4Invert(viewMatrix, nil)
        render(viewMatrix)
    }
}
